[GTK][WPE] Add provision to enable / disable websecurity

pgorszkowski-igalia opened the pull request:
#### 88ecd4bb2c647710dc440176e6642a39cbb51f0d
<pre>
[GTK][WPE] Add provision to enable / disable websecurity
<a href="https://bugs.webkit.org/show_bug.cgi?id=246606">https://bugs.webkit.org/show_bug.cgi?id=246606</a>

Reviewed by Carlos Garcia Campos.

* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webKitSettingsSetProperty):
(webKitSettingsGetProperty):
(webkit_settings_class_init):
(webkit_settings_get_enable_websecurity):
(webkit_settings_set_enable_websecurity):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettings):

Canonical link: <a href="https://commits.webkit.org/258707@main">https://commits.webkit.org/258707@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7d5845c4bb23844b2269c3698e02ace6eb525d7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100019 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9190 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33100 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109366 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169600 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104018 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10070 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86544 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92466 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107265 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105787 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/7609 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90892 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34327 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/89516 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22285 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2982 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23800 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2941 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/46172 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9075 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43281 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5986 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4791 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aperezdc commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -1581,6 +1588,20 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+    /**
+     * WebKitSettings:enable-websecurity:
+     *
+     * Enable or disable support for Web Security on pages.
```
This needs a comment here saying what exactly gets disabled (cross-domain checks, CORS, CSP, maybe others?) and a big fat scary warning telling people that they **will be hacked** if they navigate the open Internet with this setting disabled.

pgorszkowski-igalia replied:
I added some additional comment and warning but any suggestions are welcome

aperezdc reviewed (commented):
Patch looks fine but the documentation needs to be scary, very scary 👻 🎃

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -3944,3 +3969,44 @@ void webkit_settings_set_media_content_types_requiring_hardware_support(WebKitSe
     g_object_notify_by_pspec(G_OBJECT(settings), sObjProperties[PROP_MEDIA_CONTENT_TYPES_REQUIRING_HARDWARE_SUPPORT]);
 }
+
+
```
```suggestion
```

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -1581,6 +1588,24 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+     *
+     * Enable or disable support for Web Security on pages.
+     *
+     * This setting is enabled by default.
```
No need to mention this: the default value is part of the introspection metadata.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -1581,6 +1588,24 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+     *
+     * This setting is enabled by default.
+     * WARNING: Disabling Web Security on pages means disabling the same-origin policy which is 
+     * very dangerous and you do it on your own risk.
```
```suggestion
     * This setting disables the same-origin policy, allowing every website full control over
     * all other websites. This is for use in special environments where you wish to disable
     * all security and allow websites to hack each other. It is impossible to use this setting
     * securely.
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -1581,6 +1588,25 @@ static void webkit_settings_class_init(WebKitSettingsClass* klass)
+     * all security and allow websites to hack each other. It is impossible to use this setting
+     * securely.
+     *
+     * Since: 2.38
```
```suggestion
     * Since: 2.40
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -172,6 +172,7 @@ enum {
     PROP_ENABLE_MEDIA,
     PROP_MEDIA_CONTENT_TYPES_REQUIRING_HARDWARE_SUPPORT,
     PROP_ENABLE_WEBRTC,
+    PROP_ENABLE_WEBSECURITY,
```
WEB_SECURITY? same for the property name and getter/setter

carlosgcampos replied:
I know all other settings are `enable-foo`, but in this case I think I would use `disable-web-security` to make even more explicit that this setting is to disable security.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -3944,3 +3970,42 @@ void webkit_settings_set_media_content_types_requiring_hardware_support(WebKitSe
+ *
+ * Returns: %TRUE If websecurity support is enabled or %FALSE otherwise.
+ *
+ * Since: 2.38
```
```suggestion
 * Since: 2.40
```

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
```diff
@@ -3944,3 +3970,42 @@ void webkit_settings_set_media_content_types_requiring_hardware_support(WebKitSe
+ *
+ * Set the #WebKitSettings:enable-websecurity property.
+ *
+ * Since: 2.38
```
```suggestion
 * Since: 2.40
```

carlosgcampos reviewed (changes requested):
I would add another test to check the effect of changing the setting.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1934,6 +1934,54 @@ static void testWebViewWebExtensionMode(WebViewTest* test, gconstpointer)
+    auto waitForFooChanged = [&test]() {
+        GUniqueOutPtr<GError> error;
+        WebKitJavascriptResult* result;
+        JSCValue* jscvalue;
```
Move these to where they are assigned.

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1934,6 +1934,54 @@ static void testWebViewWebExtensionMode(WebViewTest* test, gconstpointer)
+        GUniqueOutPtr<GError> error;
+        WebKitJavascriptResult* result;
+        JSCValue* jscvalue;
+        int value;
```
This could be status, instead of value.

pgorszkowski-igalia replied:
This variable is used to store the value of the 'foo' variable from javascript so IMO it can stay as value or maybe fooValue. @carlosgcampos  what is your opinion?

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1934,6 +1934,54 @@ static void testWebViewWebExtensionMode(WebViewTest* test, gconstpointer)
+            g_assert_no_error(error.get());
+            jscvalue = webkit_javascript_result_get_js_value(result);
+            value = jsc_value_to_int32(jscvalue);
+            webkit_javascript_result_unref(result);
```
You don't own this result, it's owned by WebViewTest

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1934,6 +1934,54 @@ static void testWebViewWebExtensionMode(WebViewTest* test, gconstpointer)
+            jscvalue = webkit_javascript_result_get_js_value(result);
+            value = jsc_value_to_int32(jscvalue);
+            webkit_javascript_result_unref(result);
+        } while (!value);
```
This is a bit confusing, I guess this is 0 when the JSCValue is undefined. I would explicitly check if value is undefined instead.

pgorszkowski-igalia replied:
jscvalue is 0 if the fetch has not finished yet(initial value of 'foo'). If  fetch is finished, then 'foo' is set to `response.status` in case of success and to -1 in case of any problems.
<script>let foo = 0; fetch('foo://bar/baz').then(response => { foo = response.status; }).catch(err => { foo = -1; });</script>

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitWebView.cpp:
```diff
@@ -1934,6 +1934,54 @@ static void testWebViewWebExtensionMode(WebViewTest* test, gconstpointer)
+    // Request is not allowed, foo should be -1.
+    webkit_web_view_load_html(test->m_webView, html, "http://example.com");
+    test->waitUntilLoadFinished();
+    g_assert_cmpint(waitForFooChanged(), ==, -1);
```
I don't think we need to check this again, since disabling web security when already disabled did nothing.

carlosgcampos reviewed (approved):
Looks good to me now, but I would change a few things in the test.

pgorszkowski-igalia committed:
[GTK][WPE] Add provision to enable / disable websecurity
https://bugs.webkit.org/show_bug.cgi?id=246606

Reviewed by Carlos Garcia Campos.

* Source/WebKit/UIProcess/API/glib/WebKitSettings.cpp:
(webKitSettingsSetProperty):
(webKitSettingsGetProperty):
(webkit_settings_class_init):
(webkit_settings_get_enable_websecurity):
(webkit_settings_set_enable_websecurity):
* Source/WebKit/UIProcess/API/glib/WebKitSettings.h.in:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestWebKitSettings.cpp:
(testWebKitSettings):

Canonical link: https://commits.webkit.org/258707@main
